ldms/src/core/ldms_xprt.c

```diff
@@ -2108,38 +2108,77 @@ void __ldms_xprt_conn_msg_init(ldms_t _x, struct ldms_conn_msg *msg)
 
 void __ldms_xprt_init(struct ldms_xprt *x, const char *name,
 					ldms_log_fn_t log_fn);
-static char __remote_name[256];
-static char __local_name[256];
+
+pthread_mutex_t xprt_names_lock = PTHREAD_MUTEX_INITIALIZER;
+static char __remote_name[LDMS_XPRT_NAME_LEN];
+static char __remote_port[LDMS_XPRT_NAME_LEN];
+static char __local_name[LDMS_XPRT_NAME_LEN];
+static char __local_port[LDMS_XPRT_NAME_LEN];
 static const char *__names[] = {
 	__local_name,
-	__remote_name
+	__local_port,
+	__remote_name,
+	__remote_port
 };
 #define LDMS_XPRT_LCL_NAME	0
-#define LDMS_XPRT_RMT_NAME	1
-const char **ldms_xprt_names(ldms_t x)
+#define LDMS_XPRT_LCL_PORT	1
+#define LDMS_XPRT_RMT_NAME	2
+#define LDMS_XPRT_RMT_PORT	3
+const char **__ldms_xprt_names(ldms_t x, int flags)
 {
 	struct sockaddr lcl, rmt;
 	socklen_t xlen = sizeof(lcl);
 	zap_err_t zerr;
 	memset(&rmt, 0, sizeof(rmt));
 	memset(&lcl, 0, sizeof(rmt));
-	__remote_name[0] = '\0';
-	__local_name[0] = '\0';
+	__remote_name[0] = __remote_port[0] = '\0';
+	__local_name[0] = __local_port[0] = '\0';
 	zerr = zap_get_name(x->zap_ep,
 			    (struct sockaddr *)&lcl,
 			    (struct sockaddr *)&rmt, &xlen);
 	if (zerr)
 		goto out;
 	(void)getnameinfo(&rmt, xlen, __remote_name, sizeof(__remote_name),
-			  NULL, 0, NI_NUMERICHOST);
+			  __remote_port, sizeof(__remote_port), flags);
 	(void)getnameinfo(&lcl, xlen, __local_name, sizeof(__local_name),
-			  NULL, 0, NI_NUMERICHOST);
+			  __local_port, sizeof(__local_port), flags);
 out:
 	__names[0] = __local_name;
-	__names[1] = __remote_name;
+	__names[1] = __local_port;
+	__names[2] = __remote_name;
+	__names[3] = __remote_port;
 	return __names;
 }
 
+void ldms_xprt_names(ldms_t x, char *lcl_name, size_t lcl_name_sz,
+				char *lcl_port, size_t lcl_port_sz,
+				char *rem_name, size_t rem_name_sz,
+				char *rem_port, size_t rem_port_sz,
+				int flags)
+{
+	const char **names;
+	pthread_mutex_lock(&xprt_names_lock);
+	names = __ldms_xprt_names(x, flags);
```

Comment:
I didn't remove the static buffer because it makes the call to ```__ldms_xprt_names()``` simple inside ```ldms_xprt.c``` such as https://github.com/ovis-hpc/ovis/blob/22f55dcbd302194fcbf12bb2dee1f4694864e701/ldms/src/core/ldms_xprt.c#L2191.

I think I should remove the static buffer, the lock, and the internal function ```__ldms_xprt_names()``` all together because the function isn't thread-safe. The caller must provide buffers to receive the address names. What do you think?

Reply:
I didn't remove the static buffer because it makes the call to ```__ldms_xprt_names()``` simple inside ```ldms_xprt.c``` such as https://github.com/ovis-hpc/ovis/blob/22f55dcbd302194fcbf12bb2dee1f4694864e701/ldms/src/core/ldms_xprt.c#L2191.

I think I should remove the static buffer, the lock, and the internal function ```__ldms_xprt_names()``` all together because the function isn't thread-safe. The caller must provide buffers to receive the address names. What do you think?

Reply:
Because you call the function inside a mutex, and copy the data out of the static buffers into the user buffers, I think ldms_xprt_names _is_ thread save. But the point is that if you use the caller provided buffers you don't need any of that, you don't need the  LDMS_XPRT_NAME_LEN  limit, and the data doesn't get copied twice. But maybe I'm missing something?

But otherwise, yes I agree that it would be nice if we could use use the caller's buffers; avoid the extra static buffers, copy and mutex.